File: plotting/plot_grouping.py
from pathlib import Path
import re


def generate_grouped_plots(
        df,
        group_by,
        plot_function,
        output_directory,
        plot_name,
        **kwargs
):

    """
    Generate one plot per group.

    Parameters
    ----------
    df : pandas.DataFrame
        Input dataframe

    group_by : str
        Column used for grouping

    plot_function : function
        Plotting function to execute

    output_directory : str
        Folder where plots are saved

    plot_name : str
        Base name for output files

    kwargs :
        Additional plotting arguments

    Returns
    -------
    list of dictionaries
        Plot metadata for manifest
    """

    output_directory = Path(output_directory)

    output_directory.mkdir(
        exist_ok=True,
        parents=True
    )


    plot_outputs = []


    #
    # Determine groups
    #

    groups = sorted(
        df[group_by]
        .dropna()
        .unique()
    )


    #
    # Generate each plot
    #

    for group in groups:


        #
        # Filter dataframe
        #

        subset = df[
            df[group_by] == group
        ]


        #
        # Make safe filename
        #

        safe_group = re.sub(
            r"[^a-zA-Z0-9_-]",
            "_",
            str(group)
        )


        output_file = (
            output_directory /
            f"{plot_name}_{safe_group}.png"
        )


        #
        # Generate plot
        #

        plot_function(

            dataframe=subset,

            output_file=str(output_file),

            title=f"{plot_name}: {group}",

            **kwargs

        )


        #
        # Record output
        #

        plot_outputs.append(

            {

                "group": str(group),

                "file": str(output_file),

                "rows_used": len(subset)

            }

        )


    return plot_outputs

File: plotting/plot_grouping.R
############################################################
# grouped_plots.R
#
# Generic grouped plotting functions
#
# Called by:
#     plot_workflow.R
#
# Depends on:
#     basic_plots.R
#
############################################################

library(dplyr)

############################################################
# Generate one plot for each group
############################################################

generate_grouped_plots <- function(
    dataframe,
    group_by,
    plot_function,
    mapping,
    title,
    output_directory
) {

    if (!dir.exists(output_directory)) {
        dir.create(output_directory, recursive = TRUE, showWarnings = FALSE)
    }

    groups <- sort(unique(dataframe[[group_by]]))
    groups <- groups[!is.na(groups)]

    output_files <- character()

    for (group in groups) {
        subset_df <- dataframe %>%
            filter(.data[[group_by]] == group)

        safe_group <- gsub("[^A-Za-z0-9_-]", "_", as.character(group))

        output_file <- file.path(
            output_directory,
            paste0(safe_group, ".png")
        )

        group_title <- paste(title, "-", group)

        plot_function(
            dataframe = subset_df,
            mapping = mapping,
            title = group_title,
            output_directory = output_directory,
            output_file = output_file
        )

        output_files <- c(output_files, output_file)
    }

    return(output_files)
}